### PR TITLE
DataTexture: needsUpdate true by default

### PR DESCRIPTION
**Pull request by EliasHasle**
Avoids a lot of head-scratching.

### CODE REVIEW

**Comment by Mugen87** [comment]
If you want to make this change, you should also clean up places like the following:

https://github.com/mrdoob/three.js/blob/10ab11cb1e4646a2b661f99c680634406189f601/src/renderers/WebGLRenderer.js#L1814-L1815

Otherwise `needsUpdate` is unnecessarily set which increases the internal `version` counter for no reasons.

File: src/textures/DataTexture.js [hunk 1]
```diff
@@ -18,6 +18,11 @@ function DataTexture( data, width, height, format, type, mapping, wrapS, wrapT,
 	this.flipY = false;
 	this.unpackAlignment = 1;
 
+	if ( data ) {
+
+		this.needsUpdate = true;
+
+	}
```
Comment (Mugen87) [inline src/textures/DataTexture.js]: You need an empty line after this bracket. Otherwise Travis won't be happy.

**Comment by EliasHasle** [comment]
@Mugen87 Thanks for your review. Do you otherwise approve of the change? (I don't have the best experience lately with investing time in PRs.)

**Comment by Mugen87** [comment]
I'm not sure, the change feels a bit hacky. I leave it up to @mrdoob .

**Comment by EliasHasle** [comment]
Hm, maybe `width` and `height` must be required too, before automatically setting `needsUpdate`.

The problem I am trying to solve is that a user (me) creates a DataTexture and tries to use it, only to get a big black rectangle, and then literally spends hours scratching his (my) head in search of a solution. This has happened (to me) more than once. I noticed that the example code in the docs remembers to set the flag, but as always it was a surprise to see that it is required before the first use.

**Comment by gkjohnson** [comment]
I've been confused by the initial setting of `needsUpdate` to false for awhile and have lost time before trying to understand why my texture isn't working only to realize I forgot to set `needsUpdate` so if there's a way to initialize it to true I'm all for this.

Does `DataTexture` break if needsUpdate is _always_ initialized to true even when there's no data? The proposed solution doesn't handle the following case (which is how I initialize my objects typically):

```js
const tex = new DataTexture();
tex.image.data = new Float32Array(256**2);
tex.image.width = 256;
tex.image.height = 256;

// set other fields
```

**Comment by WestLangley** [comment]
The app needs to be able to control when `.needsUpdate` is set.

For example:

https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/HDRCubeTextureLoader.js#L101-L106

**Comment by mrdoob** [comment]
What's the worst that can happen if we just set `this.needsUpdate = true` by default?

Something safer could be this:

```js
var default = new Uint8Array( 4 );

function DataTexture( data, width, height, format, type, mapping, wrapS, wrapT, magFilter, minFilter, anisotropy, encoding ) {

	Texture.call( this, null, mapping, wrapS, wrapT, magFilter, minFilter, format, type, anisotropy, encoding );

	this.image = { data: data || default, width: width || 1, height: height || 1 };

	this.magFilter = magFilter !== undefined ? magFilter : NearestFilter;
	this.minFilter = minFilter !== undefined ? minFilter : NearestFilter;

	this.generateMipmaps = false;
	this.flipY = false;
	this.unpackAlignment = 1;

	this.needsUpdate = true;

}
```

**Comment by EliasHasle** [comment]
According to the WebGL spec, the datatexture `pixels` field can be null, so there should in principle not be a need for a dummy texture.

**Comment by mrdoob** [comment]
> I haven't tested it, but how about
> ```js
>	this.image = { data: data || null, width: width || 1, height: height || 1 };
> ```

I think that looks good to me.

Do you want to amend this PR so we can merge it now and have some time to test it?

**Comment by mrdoob** [comment]
Oh, you did already.

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
It would be nice to identify these cases and do some clean up:

> If you want to make this change, you should also clean up places like the following:
>
> https://github.com/mrdoob/three.js/blob/10ab11cb1e4646a2b661f99c680634406189f601/src/renderers/WebGLRenderer.js#L1814-L1815

Not sure what the best way would be... Maybe by adding this code [here](https://github.com/mrdoob/three.js/blob/dev/src/textures/Texture.js#L326)?

```js
console.error( this.uuid, this.version );
```

**Comment by mrdoob** [comment]
Okay, I think I did all of them. https://github.com/mrdoob/three.js/commit/d3a37f37d395010111f384fe0e9f73c072ad7384

**Comment by EliasHasle** [comment]
> Okay, I think I did all of them. [d3a37f3](https://github.com/mrdoob/three.js/commit/d3a37f37d395010111f384fe0e9f73c072ad7384)

Change doc too during the testing phase? https://github.com/mrdoob/three.js/blob/f5780f2a165ef8d0e11c4755b586db43d45dd551/docs/api/en/textures/DataTexture.html#L60

**Comment by mrdoob** [comment]
Oh yes, would you like to do a PR?

**Comment by Mugen87** [comment]
This change is actually problematic when we want to use `texStorage2D()` with data textures.

Data textures are now treated like canvas texture with the exception that the data source parameters are not mandatory. 

So if you create a data texture without parameters, the texture is marked as "ready for upload" although no data are defined. That is no problem with `texImage2D()` however `texStorage2D()` has to allocated the buffer for the actual texture dimensions. Right now, it would allocate a 1x1 texture and `texSubImage2D()` tries to upload `null` which is also not valid, too.

For a clean implementation I suggest to treat data textures like normal textures and do not set `needsUpdate` to `true` in the ctor. The version has to be zero in its initial state. That means this PR has to be reverted. 

Alternatively, the first three parameters have to be mandatory. That means however `DataTextureLoader` has to be refactored and potentially a lot of user code. At least it's a bigger migration task than asking to set `needsUpdate` to `true` again when manually creating a data texture.

**Comment by WestLangley** [comment]
>That means this PR has to be reverted.

Right, as I said above, the app controls when the data is ready.

**Pull request merged**